mypy/checker.py

```diff
@@ -5019,21 +5019,65 @@ def visit_if_stmt(self, s: IfStmt) -> None:
 
                 if_map, else_map = self.find_isinstance_check(e)
 
-                # XXX Issue a warning if condition is always False?
+                self._visit_if_stmt_redundant_expr_helper(
+                    stmt=s, expr=e, body=b, if_map=if_map, else_map=else_map
+                )
+
                 with self.binder.frame_context(can_skip=True, fall_through=2):
                     self.push_type_map(if_map, from_assignment=False)
                     self.accept(b)
 
-                # XXX Issue a warning if condition is always True?
                 self.push_type_map(else_map, from_assignment=False)
 
             with self.binder.frame_context(can_skip=False, fall_through=2):
                 if s.else_body:
                     self.accept(s.else_body)
 
+    def _visit_if_stmt_redundant_expr_helper(
+        self, stmt: IfStmt, expr: Expression, body: Block, if_map: TypeMap, else_map: TypeMap
+    ) -> None:
+        """Emits `redundant-expr` errors for if statements that are always true or always false.
+
+        We try to avoid emitting such errors if the redundancy seems to be intended as part of
+        dynamic type or exhaustiveness checking (risking to miss some unintended redundant if
+        statements).
+        """
+
+        if codes.REDUNDANT_EXPR not in self.options.enabled_error_codes:
+            return
+        if refers_to_fullname(expr, ("typing.TYPE_CHECKING", "typing_extensions.TYPE_CHECKING")):
+            return
+
+        def _filter(body: Block | None) -> bool:
+            if body is None:
+                return False
+            s = body.body[0]
+            if isinstance(s, AssertStmt) and is_false_literal(s.expr):
+                return True
+            if isinstance(s, RaiseStmt):
+                return True
+            elif isinstance(s, ExpressionStmt) and isinstance(s.expr, CallExpr):
+                with self.expr_checker.msg.filter_errors(filter_revealed_type=True):
+                    typ = self.expr_checker.accept(
+                        s.expr, allow_none_return=True, always_allow_any=True
+                    )
+                if isinstance(get_proper_type(typ), UninhabitedType):
+                    return True
+            return False
```

Comment:
Yeah as you point out this seems to duplicate `is_noop_for_reachability`. Why not just call that? I think probably `pass` and `...` as bodies are alright. (and any reasoning against them IMO also stands for disabling reachability warnings on them :^) )